src/box/applier.cc

```diff
@@ -1936,6 +1936,18 @@ applier_thread_data_destroy(struct applier *applier)
 	msg.applier = applier;
 	cbus_call(&thread->thread_pipe, &thread->tx_pipe, &msg.base,
 		  applier_thread_detach_applier);
+	ERROR_INJECT(ERRINJ_APPLIER_DESTROY_DELAY, {
+		struct errinj *inj = errinj(ERRINJ_WAL_IO, ERRINJ_BOOL);
+		if (inj->bparam) {
+			inj->bparam = false;
+			say_warn("applier failure destroy is delayed");
+			ERROR_INJECT_YIELD(ERRINJ_APPLIER_FAIL_DESTROY_DELAY);
```

Comment:
These three error injections are extremely interconnected.

Maybe move the ERRINJ_APPLIER_FAIL_DESTROY_DELAY here?
https://github.com/tarantool/tarantool/blob/dfe79b9aa30e9aad9b105fc559df7a9f08b8fe5b/src/box/applier.cc#L2229-L2232

Reply:
> Or maybe even better, leave APPLIER_FAIL_DESTROY_DELAY in applier_thread_data_destroy, but move APPLIER_DESTROY_DELAY to the place where FiberIsCancelled is caught.

`applier_thread_data_destroy` works for all appliers, it's really difficult to manipulate connected error injections, which yields on different "levels" of applier destruction. Unfortunately, I couldn't make this work.  Here's the problem.

Changes you proposed can indeed reproduce the problem, but it's not capable of testing the patch. As soon as `APPLIER_FAIL_DESTROY_DELAY` stops yielding, triggers're run, assertion fails. After applying patch (src/box/replication.cc) the test must end successfully. But it won't because all other appliers should be joined successfully, but they still yields. So we must set `APPLIER_DESTROY_DELAY` manually to `false` right after `APPLIER_FAIL_DESTROY_DELAY`.

These two error injections are still pretty connected, but it's much better now. They're both in `applier_f` and look pretty consistent.

Reply:
Now error injections are not connected at all. But one of them just output to log some info